viabel/optimization.py

```diff
@@ -692,114 +802,56 @@ def optimize(self, K_max, objective, init_param):
-        grad_history = None
-        descent_dir_history = None
-        iterate_average_k_history = None
-        iterate_average_history = None
-        ess_and_mcse_k_history = None
-        ess_history = None
-        mcse_history = None
-        final_mcse_history = []
-        SKL_history = []
-        learning_rate_hist = []
-        kappa_hist = []
-        c_hist = []
-        c_sample_hist = []
-        kappa_sample_hist = []
-        k_new_hist = [0]
-        k_conv_hist = []
-        k_Rhat_hist = []
+        history = defaultdict(list)
+        history['iterate_average_curr_hist'].append(iterate_average_curr)
+        history['k_mcse'].append(0)
         stopped = False
-        convg_iters_hist = []
-        predicted_iters_hist = []
-        stopping_crt = []
-        k_stopped_final_hist = []
-        LR = [] #learning rate for WLS regression 
-        CI = [] #Converged Iterations for WLS regression
         try:
             while not stopped:
-                K_max -= (k_new +1)
+                K_max -= (k_new + 1)
                 iterate_average_prev = iterate_average_curr
-                if k==0 and self._init_rmsprop:
+                if k == 0 and self._init_rmsprop: #Using RMSProp optimization initially if specified
                     rmsprop = RMSProp(learning_rate=sgo._learning_rate, diagnostics=diagnostics)
                     faso = FASO(sgo = rmsprop)
                     opt = faso.optimize(K_max, objective, iterate_average_curr)
                 else:
                     opt = super().optimize(K_max, objective, iterate_average_curr)
-                if opt['k_stopped'] is not None and k!=0: #removing the number of convergence iterations of initial learning rate
+                if opt['k_stopped'] is not None and k != 0: #removing the number of convergence iterations of initial learning rate
                     convg_iters = opt['k_stopped']
-                    convg_iters_hist.append(convg_iters)
-                    CI.append(convg_iters)
+                    history['conv_iters_hist'].append(convg_iters)
+                    # CI.append(convg_iters)
                 iterate_average_curr = opt['opt_param']
-                iterate_average_curr_hist.append(iterate_average_curr)
+                history['iterate_average_curr_hist'].append(iterate_average_curr)
                 k_new = opt['k_stopped']
                 
-                if opt['k_Rhat'] is not None and k_new is not None:
-                    k_Rhat_hist.append(opt['k_Rhat']+k_add)
-                else:
-                    k_Rhat_hist.append(opt['k_Rhat'])
-                    
-                if opt['k_conv'] is not None and k_new is not None:
-                    k_conv_hist.append(opt['k_conv']+k_add)
-                else:
-                    k_conv_hist.append(opt['k_conv'])
+                history['k_Rhat'].append(opt['k_Rhat'] + k_add if opt['k_Rhat'] is not None and k_new is not None else opt['k_Rhat'])   
```

Comment:
Why are you checking whether `k_new is not None` here? The logic for this check is not super clear to me so it would be helpful to add a brief comment explaining what's going on